### ISSUE TITLE
Frontend compiler plugin should throw a warning for the possible wrong use of constructs.

### ISSUE THREAD

**Issue by angelacorte** [Collektive/collektive#575]
The following use of the `repeat`( /`repeating` ) construct with a nested `neighboring`, in some cases should be highlighted as a warning from the frontend compiler plugin.

Eg:
``` kotlin
repeat(initial){ it ->
    val newValue = //same of `it`
    neighboring(newValue).operation //field reduction
}
```
In the case above, 
the compiler should throw a warning that suggest to use the `share` construct instead of nesting `repeat` and `neighboring`.
If what you share with your neighbors (using `neighboring`), and it depends on the state of the `repeat`, 
is different from what you return as the last value of the `repeat`, then the entire block can be replaced with a `share`.

``` kotlin
repeat(initial){ it ->
    val newValue = //same of `it`
    neighboring(otherValue)
    newValue
}
```
Different is the case in which a different value is shared among neighbors, in this case it is ok to use the nested constructs.

You may also find a pattern for this use.

**Comment by nicolasfara** [comment]
@FreshMag Could you check if this scenario is already covered in the frontend plugin?
If not, could you integrate it into the plugin?

**Comment by nicolasfara** [comment]
@FreshMag could you give this a spin?

**Comment by FreshMag** [comment]
Yes, I'm inspecting this right now. Can you give me an example of substitution using `share`? For example in this case:

```kotlin
evolve(0) { value ->
    val newValue = value + 1
    neighboring(newValue).max(0)
}
``` 
how would you use `share` instead of `evolve`?

**Comment by angelacorte** [comment]
@FreshMag 
> how would you use `share` instead of `evolve`?

Is the combination of `evolve` and `neighboring` that should be replaced with `share`.
`evolve` models the state evolution of the device over time, while `neighboring` enables access to the values of neighbors and send them information.
`share` combines these two constructs, modeling the space-time evolution.

In your example it should be:
```kotlin
share(0) { value ->
    (value + 1).max(0)
}
```

**Comment by FreshMag** [comment]
@angelacorte I have a question regarding this check. In the case of `evolving`, the expression we need to check is the one contained in the `yielding` block correct?

For example in this case:

```kotlin
evolving(0) { value ->
    neighboring(value)
    value.yielding { value }
}
```

This should not raise a warning since even though the `neighboring` construct depends on the result of `evolving`, the expression shared by neighboring and the one return by the `yielding` block are the same.

This one, instead:

```kotlin
evolving(0) { 
    val newValue = it + 1
    neighboring(newValue).max(0).yielding { "test" }
}
```  

should raise a warning since the expressions are different.

Please correct me if I'm wrong

**Comment by angelacorte** [comment]
@FreshMag 
> In the case of `evolving`, the expression we need to check is the one contained in the `yielding` block correct?
> 
> For example in this case:
> 
> ```kotlin
> evolving(0) { value ->
>     neighboring(value)
>     value.yielding { value }
> }
> ```
> 
> This should not raise a warning since even though the `neighboring` construct depends on the result of `evolving`, the expression shared by neighboring and the one return by the `yielding` block are the same.

> This one, instead:
> 
> ```kotlin
> evolving(0) { 
>     val newValue = it + 1
>     neighboring(newValue).max(0).yielding { "test" }
> }
> ```
> 
> should raise a warning since the expressions are different.
> 
> Please correct me if I'm wrong

The expression to check is the one that should be sent to the other devices, the `yielding` is used to return a different value from the one sent (see [this doc](https://github.com/Collektive/collektive/blob/90195bd99cc4b6ea68d636a327d0f1464e976d75/dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/YieldSupport.kt#L25) for better examples).
Therefore, the expression to check is **not** the one inside the `yielding` block (and should throw the warning).
